File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementSerializationMethodsCorrectly.vb
Imports System
Imports System.Runtime.Serialization

Namespace Tests.Diagnostics
    <Serializable>
    Public Class Foo
        <OnSerializing>
        Public Sub OnSerializing(ByVal context As StreamingContext) ' Noncompliant {{Make this method 'Private'.}}
'                  ^^^^^^^^^^^^^
        End Sub

        <OnSerialized>
        Private Function OnSerialized(ByVal context As StreamingContext) As Integer ' Noncompliant {{Make this method a 'Sub' not a 'Function'.}}
        End Function

        <OnDeserializing>
        Private Sub OnDeserializing() ' Noncompliant  {{Make this method have a single parameter of type 'StreamingContext'.}}
        End Sub

        <OnDeserialized>
        Private Sub OnDeserialized(ByVal context As StreamingContext, ByVal str As String) ' Noncompliant {{Make this method have a single parameter of type 'StreamingContext'.}}
        End Sub

        Private Sub OnDeserialized2(ByVal context As StreamingContext, ByVal str As String)
        End Sub

        <OnDeserialized>
        Private Sub OnDeserialized(Of T)(ByVal context As StreamingContext) ' Noncompliant {{Make this method have no type parameters.}}
        End Sub

        <OnDeserializing>
        Public Function OnDeserializing2(ByVal context As StreamingContext) As Integer ' Noncompliant {{Make this method 'Private' and a 'Sub' not a 'Function'.}}
            Throw New NotImplementedException()
        End Function

        <OnDeserializing>
        Public Sub OnDeserializing3() ' Noncompliant {{Make this method 'Private' and have a single parameter of type 'StreamingContext'.}}
            Throw New NotImplementedException()
        End Sub

        <OnDeserializing>
        Private Function OnDeserializing4() As Integer ' Noncompliant {{Make this method a 'Sub' not a 'Function' and have a single parameter of type 'StreamingContext'.}}
            Throw New NotImplementedException()
        End Function

        <OnDeserializing>
        Public Function OnDeserializing5(Of T)() As Integer ' Noncompliant {{Make this method 'Private', a 'Sub' not a 'Function', have no type parameters and have a single parameter of type 'StreamingContext'.}}
            Throw New NotImplementedException()
        End Function

        <OnDeserializing>
        Public Function () As String ' Noncompliant ' Error [BC30203]
            Throw New NotImplementedException()
        End Function
    End Class
End Namespace

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementSerializationMethodsCorrectly.cs
using System;
using System.Runtime.Serialization;

namespace Tests.Diagnostics
{
    [Serializable]
    public class Foo
    {
        [OnSerializing]
        public void OnSerializing(StreamingContext context) { } // Noncompliant {{Make this method 'private'.}}
//                  ^^^^^^^^^^^^^

        [OnSerialized]
        int OnSerialized(StreamingContext context) { return 1; } // Noncompliant {{Make this method return 'void'.}}

        [OnDeserializing]
        void OnDeserializing() { } // Noncompliant  {{Make this method have a single parameter of type 'StreamingContext'.}}

        [OnDeserialized]
        void OnDeserialized(StreamingContext context, string str) { } // Noncompliant {{Make this method have a single parameter of type 'StreamingContext'.}}

        void OnDeserialized2(StreamingContext context, string str) { } // Compliant

        [OnDeserialized]
        void OnDeserialized<T>(StreamingContext context) { } // Noncompliant {{Make this method have no type parameters.}}

        [OnDeserializing]
        public int OnDeserializing2(StreamingContext context) { throw new NotImplementedException(); } // Noncompliant {{Make this method 'private' and return 'void'.}}

        [OnDeserializing]
        public void OnDeserializing3() { throw new NotImplementedException(); } // Noncompliant {{Make this method 'private' and have a single parameter of type 'StreamingContext'.}}

        [OnDeserializing]
        int OnDeserializing4() { throw new NotImplementedException(); } // Noncompliant {{Make this method return 'void' and have a single parameter of type 'StreamingContext'.}}

        [OnDeserializing]
        public int OnDeserializing5<T>() { throw new NotImplementedException(); } // Noncompliant {{Make this method 'private', return 'void', have no type parameters and have a single parameter of type 'StreamingContext'.}}
    }
}
